File: java/leetcode/ValidSquare.java
//Leetcode problem 367 Valid Perfect Square
//Solution written by Xuqiang Fang on  16 April, 2018

class Solution{
    public boolean isPerfectSquare(int num) {
        if(num == 1)
            return true;
        int l = 1;
        int r = num / 2;
        int m = l + (r - l) / 2;
        while(l < r){
            if((l == num / l && num % l == 0) || (r == num / r && num % r == 0))
                return true;
            m = l + (r - l) / 2;
            if(m == num / m && num % m == 0)
                return true;
            else if(m < num / m + 1)
                l = m + 1;
            else
                r = m - 1;
        }
        return false;
    }	
}

public class ValidSquare{
	public static void main(String[] args){
		Solution s = new Solution();
        for(int i=1; i<=10000; i++){
            if(s.isPerfectSquare(i))
                System.out.println(i);
        }
        System.out.println(s.isPerfectSquare(Integer.valueOf(args[0])));
	}
}

File: cpp/leetcode/ValidSquare.cpp
//Leetcode Problem No 593 Valid Square
//Solution written by Xuqiang Fang on 9 Nov, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution {
public:
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        if(valid(p1, p2, p3, p4) || valid(p1, p3, p2, p4) || valid(p1, p4, p2, p3) )
            return true;
        return false;
    }
private:
    bool valid(vector<int>& p, vector<int>& q, vector<int>& r, vector<int>& s){
        if(p[0] == q[0]){
            return r[1] == s[1] && distance(p,r) == distance(q,r) && distance(p,s) == distance(q,s) && distance(p,q) == distance(r,s) && distance(p,q) != 0;
        }
        else if(r[0] == s[0]){
            return p[1] == q[1] && distance(p,r) == distance(q,r) && distance(p,s) == distance(q,s) && distance(p,q) == distance(r,s) && distance(p,q) != 0;
        }
        else{
            return (p[1]-q[1]) * (r[1]-s[1]) + (p[0]-q[0]) * (r[0]-s[0]) == 0 &&  distance(p,r) == distance(q,r) && distance(p,s) == distance(q,s) && distance(p,q) == distance(r,s) && distance(p,q) != 0;
        }
    }
    int distance(vector<int>& p, vector<int>& q){
        return (p[0] - q[0]) * (p[0]-q[0]) + (p[1]-q[1])*(p[1]-q[1]);
    }
};
int main(){
    Solution s;
    vector<int> p1{0,0}, p2{0,1}, p3{0,2}, p4{1,2};
    cout << s.validSquare(p1, p2, p3, p4) << endl;
    return 0;
}
